### ISSUE TITLE
IllegalArgumentException in 5.2.3

### ISSUE THREAD

**Issue by ToshihikoMakita** [xmlresolver/xmlresolver#184]
When I use 5.2.3 with Saxon, I got the IllegalArgumentException shown below. This prevents to upgrade Saxon to 12.5.

```
PS D:\My_Documents\XML2021\NKBJ\proj\ymc-xml-to-pdf> .\build-db2db.ps1 D:\My_Documents\XML2021\NKBJ\docs\2024-04-18-mc-data\sm\SBBVT0T\xmllist\SBBVT0T-Deployment-Flat.xml  D:\My_Documents\XML2021\NKBJ\docs\2024-04-18-mc-data\sm\SBBVT0T\xmllist\SBBVT0T-expand.xml SM C:/PROGRA~1/Java/jdk-17.0.3.1  D:\My_Documents/Java/SaxonPE12-5J/saxon-pe-12.5.jar
Start expanding entities in XML file
Input file: D:/My_Documents/XML2021/NKBJ/docs/2024-04-18-mc-data/sm/SBBVT0T/xmllist/SBBVT0T-Deployment-Flat.xml
Output file: D:/My_Documents/XML2021/NKBJ/docs/2024-04-18-mc-data/sm/SBBVT0T/xmllist/SBBVT0T-expand.xml
java.lang.IllegalArgumentException: Illegal character in path at index 2: ..\master\glossary\gls.ent
        at java.base/java.net.URI.create(URI.java:906)
        at java.base/java.net.URI.resolve(URI.java:1089)
        at org.xmlresolver.Resolver.openConnection(Resolver.java:250)
        at org.xmlresolver.Resolver.resolveEntity(Resolver.java:204)
        at net.sf.saxon.lib.CatalogResourceResolver.resolveEntity(CatalogResourceResolver.java:194)
        at net.sf.saxon.lib.EntityResolverWrappingResourceResolver.resolveEntity(EntityResolverWrappingResourceResolver.java:46)
        at java.xml/com.sun.org.apache.xerces.internal.util.EntityResolver2Wrapper.resolveEntity(EntityResolver2Wrapper.java:178)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLEntityManager.resolveEntityAsPerStax(XMLEntityManager.java:1026)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLEntityManager.startEntity(XMLEntityManager.java:1307)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDTDScannerImpl.startPE(XMLDTDScannerImpl.java:732)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDTDScannerImpl.skipSeparator(XMLDTDScannerImpl.java:2101)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDTDScannerImpl.scanDecls(XMLDTDScannerImpl.java:2064)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDTDScannerImpl.scanDTDInternalSubset(XMLDTDScannerImpl.java:356)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$DTDDriver.dispatch(XMLDocumentScannerImpl.java:1091)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$DTDDriver.next(XMLDocumentScannerImpl.java:1040)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$PrologDriver.next(XMLDocumentScannerImpl.java:917)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:112)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:542)     
        at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:889)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:825)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1224)
        at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:637)
        at net.sf.saxon.resource.ActiveSAXSource.deliver(ActiveSAXSource.java:192)
        at net.sf.saxon.resource.ActiveStreamSource.deliver(ActiveStreamSource.java:73)
        at net.sf.saxon.event.Sender.send(Sender.java:104)
        at net.sf.saxon.Configuration.buildDocumentTree(Configuration.java:4273)
        at net.sf.saxon.s9api.DocumentBuilder.build(DocumentBuilder.java:334)
        at net.sf.saxon.Transform.processFile(Transform.java:1361)
        at net.sf.saxon.Transform.doTransform(Transform.java:894)
        at net.sf.saxon.Transform.main(Transform.java:84)
Caused by: java.net.URISyntaxException: Illegal character in path at index 2: ..\master\glossary\gls.ent
        at java.base/java.net.URI$Parser.fail(URI.java:2974)
        at java.base/java.net.URI$Parser.checkChars(URI.java:3145)
        at java.base/java.net.URI$Parser.parseHierarchical(URI.java:3227)
        at java.base/java.net.URI$Parser.parse(URI.java:3186)
        at java.base/java.net.URI.<init>(URI.java:623)
        at java.base/java.net.URI.create(URI.java:904)
        ... 31 more
Fatal error during transformation: java.lang.IllegalArgumentException: Illegal character in path at index 2: ..\master\glossary\gls.ent

XSLT processing return code: 2
PS D:\My_Documents\XML2021\NKBJ\proj\ymc-xml-to-pdf>
```
We got the normal result in 4.4.0

```
PS D:\My_Documents\XML2021\NKBJ\proj\ymc-xml-to-pdf> .\build-db2db.ps1 D:\My_Documents\XML2021\NKBJ\docs\2024-04-18-mc-data\sm\SBBVT0T\xmllist\SBBVT0T-Deployment-Flat.xml  D:\My_Documents\XML2021\NKBJ\docs\2024-04-18-mc-data\sm\SBBVT0T\xmllist\SBBVT0T-expand.xml SM C:/PROGRA~1/Java/jdk-17.0.3.1  D:\My_Documents/Java/SaxonPE12-5J/saxon-pe-12.5.jar
Start expanding entities in XML file
Input file: D:/My_Documents/XML2021/NKBJ/docs/2024-04-18-mc-data/sm/SBBVT0T/xmllist/SBBVT0T-Deployment-Flat.xml
Output file: D:/My_Documents/XML2021/NKBJ/docs/2024-04-18-mc-data/sm/SBBVT0T/xmllist/SBBVT0T-expand.xml

XSLT processing return code: 0
Processing takes 0 minutes  1 seconds.

Job ended at 2024/07/08 15:55:58
```
I attached gls.ent file for your investigation.

[gls.zip](https://github.com/user-attachments/files/16134674/gls.zip)

I also attached the PS script. I used it in Windows 11.

[build-db2db.zip](https://github.com/user-attachments/files/16134685/build-db2db.zip)

Regards,

**Comment by ndw** [comment]
Where does `..\master\glossary\gls.ent` occur? Is it in an `href` attribute value or a DTD system identifier, or somewhere else? (If it's possible to send a complete test case that demonstrates the problem, that would be great.)

You say this prevents upgrading to Saxon 12.5, what version are you attempting to upgrade *from*?

The problem here is that `\` is not a legal character in URIs, although it's obviously a perfectly legal character in Windows file names. I would have expected the resolver to change that into a `/` before attempting to resolve it with `java.net.URI` but without a test case, I can only speculate about why it isn't.

**Comment by ndw** [comment]
I'm also curious, do you have additional Java customizations? I can't find anything in Saxon or XML Resolver that would print the message `Start expanding entities in XML file`

**Comment by ToshihikoMakita** [comment]
Sorry for the late reply.
The customer is planning to upgrade Saxon 12.0 to Saxon 12.5.
I attached small input file to reproduce the problem.

[SBBVT0T-Deployment-Flat-mod.zip](https://github.com/user-attachments/files/16168230/SBBVT0T-Deployment-Flat-mod.zip)

If you do need the DTD, I will negotiate with user because it is proprietary.
Regards,

**Comment by michaelhkay** [comment]
When I supply the document as a source document to Saxon Transform, I get

```
java.lang.IllegalArgumentException: java.net.URISyntaxException: Illegal character in path at index 2: ..\master\glossary\gls.ent
	at org.xmlresolver.CatalogResolver.resolveEntity(CatalogResolver.java:292)
	at org.xmlresolver.Resolver.resolveEntity(Resolver.java:199)
	at net.sf.saxon.lib.CatalogResourceResolver.resolveEntity(CatalogResourceResolver.java:194)
	at net.sf.saxon.lib.EntityResolverWrappingResourceResolver.resolveEntity(EntityResolverWrappingResourceResolver.java:46)
	at org.apache.xerces.util.EntityResolver2Wrapper.resolveEntity(Unknown Source)
	at org.apache.xerces.impl.XMLEntityManager.resolveEntity(Unknown Source)
	at org.apache.xerces.impl.XMLEntityManager.startEntity(Unknown Source)
```
So the question is, what would we expect to happen?

The XML spec says:

> System identifiers (and other XML strings meant to be used as URI references) may contain characters that, according to [[IETF RFC 3986]], must be escaped before a URI can be used to retrieve the referenced resource. The characters to be escaped are the control characters #x0 to #x1F and #x7F (most of which cannot appear in XML), space #x20, the delimiters '<' #x3C, '>' #x3E and '"' #x22, the unwise characters '{' #x7B, '}' #x7D, '|' #x7C, '\' #x5C, '^' #x5E and '`' #x60, as well as all characters above #x7F.

I doubt that escaping (that is, percent-encoding) the "unwise" character "\" as suggested here is going to be helpful; the backslash is almost certainly intended as a Windows path separator, and the only helpful action would be to convert it to "/". I guess that if that is anyone's job, it's the resolver's. This isn't strictly conformant with the XML spec, but I doubt anyone will mind.

Saxon could usefully catch the `IllegalArgumentException` and turn it into something more meaningful at the XSLT level, treating it like any other XML parsing failure.

The best course of action, of course, would be to use valid URIs, rather than relying on behaviour beyond what's mandated in the standards. I do recognize that a lot of XML software tolerates Windows filenames in contexts where URIs are expected, but it's never entirely clear, for example, how relative filenames should be resolved against absolute URIs.

**Comment by michaelhkay** [comment]
>Saxon could usefully catch the IllegalArgumentException and turn it into something more meaningful at the XSLT level, treating it like any other XML parsing failure.

Saxon does in fact catch the `IllegalArgumentException` and wrap it in a `SAXException`, which seems to me to be the right course of action.

**Comment by ToshihikoMakita** [comment]
> I do recognize that a lot of XML software tolerates Windows filenames in contexts where URIs are expected

Yes, I absolutely agree. This user has been using Adobe FrameMaker for over 20 years and it uses Windows path as URI without the slightest hesitation. I'm struggling this kind of problem because there are bunch of XML files in their CMS and I cannot fix them because they are user data.
I do expect resolver to have good solution for this kind of complex problem.

**Comment by ndw** [comment]
I think I've managed to reproduce it. It appears to be somewhere deep in Xerces:

```
  0 = {StackTraceElement@2290} "java.base/java.net.URL.<init>(URL.java:645)"
  1 = {StackTraceElement@2291} "java.base/java.net.URL.<init>(URL.java:541)"
  2 = {StackTraceElement@2292} "java.base/java.net.URL.<init>(URL.java:488)"
  3 = {StackTraceElement@2293} "org.apache.xerces.impl.XMLEntityManager.setupCurrentEntity(Unknown Source)"
  4 = {StackTraceElement@2294} "org.apache.xerces.impl.XMLEntityManager.startEntity(Unknown Source)"
  5 = {StackTraceElement@2295} "org.apache.xerces.impl.XMLEntityManager.startEntity(Unknown Source)"
  6 = {StackTraceElement@2296} "org.apache.xerces.impl.XMLDTDScannerImpl.startPE(Unknown Source)"
  7 = {StackTraceElement@2297} "org.apache.xerces.impl.XMLDTDScannerImpl.skipSeparator(Unknown Source)"
  8 = {StackTraceElement@2298} "org.apache.xerces.impl.XMLDTDScannerImpl.scanDecls(Unknown Source)"
  9 = {StackTraceElement@2299} "org.apache.xerces.impl.XMLDTDScannerImpl.scanDTDInternalSubset(Unknown Source)"
  10 = {StackTraceElement@2300} "org.apache.xerces.impl.XMLDocumentScannerImpl$DTDDispatcher.dispatch(Unknown Source)"
  11 = {StackTraceElement@2301} "org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanDocument(Unknown Source)"
  12 = {StackTraceElement@2302} "org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)"
  13 = {StackTraceElement@2303} "org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)"
  14 = {StackTraceElement@2304} "org.apache.xerces.parsers.XMLParser.parse(Unknown Source)"
  15 = {StackTraceElement@2305} "org.apache.xerces.parsers.AbstractSAXParser.parse(Unknown Source)"
  16 = {StackTraceElement@2306} "org.apache.xerces.jaxp.SAXParserImpl$JAXPSAXParser.parse(Unknown Source)"
  17 = {StackTraceElement@2307} "java.xml/org.xml.sax.helpers.XMLFilterImpl.parse(XMLFilterImpl.java:357)"
  18 = {StackTraceElement@2308} "org.xmlresolver.tools.ResolvingXMLFilter.parse(ResolvingXMLFilter.java:119)"
```

Debugging in Xerces is always a challenge. My best guess is that `setupCurrentEntity` is trying to make a URL out of the entity system identifier *before* passing it to the resolver. There's nothing I can do about that.

This doesn't explain why it works with Saxon 12.0 though.

**Comment by ToshihikoMakita** [comment]
Forgive me submitting basic question.
> $ java -jar saxon-he-12.0.jar -s:iss0184/src/SBBVT0T-Deployment-Flat-mod.xml -xsl:/Volumes/Documents/xsl/identity.xsl
In this command-line, how Saxon find xmlresolver jar file?

In my case, I invoked Saxon using following Power Shell script code. (Attached as build-db2db.zip for your reference)

[Excerpt from attached build-db2db.ps1]
```
# XMLResolver 4.4.0 .jar file path
# Reference: https://saxonica.plan.io/issues/5580

$xml_resolver_path = $ps_dir + '/xmlresolver/xmlresolver-4.6.0/lib/xmlresolver-4.6.0.jar'
# $xml_resolver_path = $ps_dir + '/xmlresolver/xmlresolver-4.4.0/lib/xmlresolver-4.4.0.jar'
# $xml_resolver_path = $ps_dir + '/xmlresolver/xmlresolver-5.2.3/lib/xmlresolver-5.2.3.jar'


# --------------------------------------------------------- #
# Expand entity in XML file                                 #
# --------------------------------------------------------- #

# Java command option & Saxon option/parameter
$java_option = " -Xmx1024m -cp $xml_resolver_path;$saxon_path; -Dxml.catalog.fixWindowsSystemIdentifiers=true"
$java_class = ' net.sf.saxon.Transform'
$java_argument = ' -xsl:' + $xsl_path + ' -s:' + $input_xml_path + ' -o:' + $output_xml_path  + ' -ea:on' + ' PRM_INPUT_FILE_URL=' + $input_xml_url
$java_cmd_param = $java_option + $java_class + $java_argument

Write-Host "Start expanding entities in XML file" -ForegroundColor Cyan -BackgroundColor Black;
Write-Host "Input file: $input_xml_path" -ForegroundColor Cyan -BackgroundColor Black;
Write-Host "Output file: $output_xml_path" -ForegroundColor Cyan -BackgroundColor Black;
if ($v -eq $TRUE){
    Write-Host "Input file url: $input_xml_url" -ForegroundColor Cyan -BackgroundColor Black;
    Write-Host "Java executable: $java_exec" -ForegroundColor Cyan -BackgroundColor Black;
    Write-Host "Java command parameter: $java_cmd_param" -ForegroundColor Cyan -BackgroundColor Black;
}

$start_date_time = Get-Date

# $java_cmd = $java_exec + $java_cmd_option
# Invoke-Expression -Command $java_cmd

$ret = [Proc.Tools.exec]::runCommand($java_exec, $java_cmd_param);

Write-Host 'XSLT processing return code:' $ret  -ForegroundColor Magenta -BackgroundColor Black
if ($ret -ne 0) {exit}

$end_date_time = Get-Date
$end_date_time_formatted = Get-FormatDateTime -dateTime $end_date_time -format 'yyyy/MM/dd HH:mm:ss'
$time_span = Get-TimeSpan -startTime $start_date_time -endTime $end_date_time
Write-Host 'Processing takes' $time_span.Minutes 'minutes ' $time_span.Seconds 'seconds.' -ForegroundColor Cyan -BackgroundColor Black
Write-Host "Job ended at $end_date_time_formatted" -ForegroundColor Cyan -BackgroundColor Black
```

[build-db2db.zip](https://github.com/user-attachments/files/16273579/build-db2db.zip)

**Comment by ToshihikoMakita** [comment]
> $ java -jar saxon-he-12.0.jar -s:iss0184/src/SBBVT0T-Deployment-Flat-mod.xml -xsl:/Volumes/Documents/xsl/identity.xsl

I'm using Saxon-PE because XSL stylesheet uses reflexive extension functions.

**Comment by ndw** [comment]
Here's my attempt to reproduce the Powershell script:

```
XML_RESOLVER_PATH=`pwd`/xmlresolver-4.6.0.jar
SAXON_PATH=`pwd`/saxon120/saxon-pe-12.0.jar

JAVA_OPTION=" -Xmx1024m -cp $XML_RESOLVER_PATH:$SAXON_PATH -Dxml.catalog.fixWindowsSystemIdentifiers=true"
JAVA_CLASS=" net.sf.saxon.Transform"
JAVA_ARG=" -xsl:identity.xsl -s:src/SBBVT0T-Deployment-Flat-mod.xml -o:out.xml -ea:on"

java $JAVA_OPTION $JAVA_CLASS $JAVA_ARG
```

That runs

```
java -Xmx1024m -cp /tmp/x/xmlresolver-4.6.0.jar:/tmp/x/saxon120/saxon-pe-12.0.jar -Dxml.catalog.fixWindowsSystemIdentifiers=true net.sf.saxon.Transform -xsl:identity.xsl -s:src/SBBVT0T-Deployment-Flat-mod.xml -o:out.xml -ea:on
```

and fails in just the same way that it did before. My source document `src/SBBVT0T-Deployment-Flat-mod.xml` is

```
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE book SYSTEM "ymcsm2-4.dtd"[
<!--
<!ENTITY G110688 SYSTEM "..\master\illustration\G110688.eps" NDATA epsi>
<!ENTITY G110689 SYSTEM "..\master\illustration\G110689.eps" NDATA epsi>
-->
<!-- Begin Glossary entities -->
<!ENTITY % gls.entities SYSTEM "..\master\glossary\gls.ent">
%gls.entities;
<!-- End Glossary entities -->
]>
<book>&test;</book>
```

my DTD, `src/ymcsm2-4.dtd` is

```
<!ELEMENT book (#PCDATA)*>
```

and my entity file, `master/glossary/gls.ent` is

```
<!ENTITY test "This is a test. This is only a test.">
```

Can you reproduce a working example using those files and with a simpler driver script?

**Comment by ToshihikoMakita** [comment]
Thank you for your reply.
I am testing more simpler one without PowerShell.

**Comment by ToshihikoMakita** [comment]
In my operating environment, at least saxon-pe-12.0.jar plus xmlresolver-4.6.0.jar seems to be working. Other combination will report java.lang.IllegalArgumentException.

I attached the test data folder: saxon-test-3-2024-07-18.zip

In this test data, the DTD file is empty. The original DTD file refers lots of entities and it is meaningless to attach them all.
Here are my screen shots. 

![2024-07-18-9](https://github.com/user-attachments/assets/ce71ffa4-dc46-4fc3-b1c9-10b95d0f3f31)

![2024-07-18-10](https://github.com/user-attachments/assets/32e0e2ba-d5be-4627-8df2-19b7ef39558f)

All of the command-lines are stored in invoke.txt.

[saxon-test-3-2024-07-18.zip](https://github.com/user-attachments/files/16283836/saxon-test-3-2024-07-18.zip)

Is this information useful for you??

**Comment by ndw** [comment]
Yes. Very. I have construct a passing case on a Windows machine. That's difficult and tedious for debugging, but it's a place to start. Thank you.

**Comment by ndw** [comment]
Right. I believe I've found and fixed the bug in (what will be 5.2.4). Next, I want to see if I can fix it in the 6.x branch.
